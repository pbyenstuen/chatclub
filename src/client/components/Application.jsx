import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import useLoader from "./useLoader";
import HeaderBar from "./HeaderBar";
import LoginPage from "./LoginPage";
import SignupPage from "./SignupPage";
import ProfilePage from "./ProfilePage";
import ChatPage from "./ChatPage";
import ChattersPage from "./ChattersPage";
import EditChatterPage from "./EditChatterPage";
import ProtectedRoute from "./ProtectedRoute";
import NotFound from "./NotFound";

const Application = ({ api }) => {
    const { data: user, reload: updateUser } = useLoader(async () => await api.auth.getUser());

    return (
        <Router>
            <>
                <HeaderBar api={api} user={user} updateUser={updateUser} />
                <Switch>
                    <ProtectedRoute
                        exact path={"/"}
                        user={user}
                        component={ChatPage}
                        api={api}
                    />
                    <ProtectedRoute
                        exact path={"/chatters"}
                        user={user}
                        component={ChattersPage}
                        api={api}
                    />
                    <ProtectedRoute
                        path={"/chatters/:id/edit"}
                        user={user}
                        component={EditChatterPage}
                        api={api}
                    />
                    <ProtectedRoute
                        path={"/profile"}
                        user={user}
                        component={ProfilePage}
                        api={api}
                    />
                    <Route path={"/login"}>
                        <LoginPage api={api} updateUser={updateUser} />
                    </Route>
                    <Route path={"/signup"}>
                        <SignupPage api={api} />
                    </Route>
                    <Route>
                        <NotFound />
                    </Route>
                </Switch>
            </>
        </Router>
    );
}

export default Application;
