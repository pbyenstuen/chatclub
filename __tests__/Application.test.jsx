import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import Application from "../src/client/components/Application";

require('react-router-dom').BrowserRouter = ({ children }) => <div>{children}</div>

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component, entries) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter initialEntries={[entries]}>{component}</MemoryRouter>, container);
    });
}

describe("Application", () => {
    it("routes to NotFound on invalid path", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "Test",
                    displayName: "Test",
                })
            }
        }

        await mountComponent(<Application api={api} />, "/nothing");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("NOT FOUND: 404");
    });

    it("can route to correct path", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "Test",
                    displayName: "Test",
                })
            }
        }

        await mountComponent(<Application api={api} />, "/login");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("SIGN IN");
    });

    it("routes to /login when not authenticated", async () => {
        const api = {
            auth: {
                getUser: () => {
                    return undefined;
                }
            }
        }

        await mountComponent(<Application api={api} />, "/profile");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("SIGN IN");
    });
});
