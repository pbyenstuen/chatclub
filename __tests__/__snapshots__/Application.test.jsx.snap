// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application can route to correct path 1`] = `"<div><div id=\\"header-bar\\"><h1>Chatclub</h1><nav><ul><li><a href=\\"/chatters\\">CHATTERS</a></li><li><a href=\\"/\\">CHAT</a></li><li><a href=\\"/profile\\">PROFILE</a></li></ul></nav><div id=\\"header-bar-user-info\\"><h3>Test</h3><button>SIGN OUT</button></div></div><div id=\\"login-container\\"><header id=\\"auth-header\\"><h2>SIGN IN</h2></header><form><div><label>Username <input type=\\"text\\" name=\\"username\\" value=\\"\\"></label></div><div><label>Password <input type=\\"password\\" name=\\"password\\" value=\\"\\"></label></div><button class=\\"submit-btn\\">SIGN IN</button></form><div id=\\"signup-options-container\\"><h3>Don't have an account?</h3><a href=\\"/signup\\">SIGN UP</a><h4>OR</h4><a href=\\"/api/auth/google\\"><button id=\\"google-btn\\">SIGN IN WITH GOOGLE</button></a></div></div></div>"`;

exports[`Application routes to /login when not authenticated 1`] = `"<div><div id=\\"header-bar\\"><h1>Chatclub</h1><nav><ul><li><a href=\\"/signup\\">SIGN UP</a></li><li><a href=\\"/login\\">SIGN IN</a></li></ul></nav></div><div id=\\"login-container\\"><header id=\\"auth-header\\"><h2>SIGN IN</h2></header><form><div><label>Username <input type=\\"text\\" name=\\"username\\" value=\\"\\"></label></div><div><label>Password <input type=\\"password\\" name=\\"password\\" value=\\"\\"></label></div><button class=\\"submit-btn\\">SIGN IN</button></form><div id=\\"signup-options-container\\"><h3>Don't have an account?</h3><a href=\\"/signup\\">SIGN UP</a><h4>OR</h4><a href=\\"/api/auth/google\\"><button id=\\"google-btn\\">SIGN IN WITH GOOGLE</button></a></div></div></div>"`;

exports[`Application routes to NotFound on invalid path 1`] = `"<div><div id=\\"header-bar\\"><h1>Chatclub</h1><nav><ul><li><a href=\\"/chatters\\">CHATTERS</a></li><li><a href=\\"/\\">CHAT</a></li><li><a href=\\"/profile\\">PROFILE</a></li></ul></nav><div id=\\"header-bar-user-info\\"><h3>Test</h3><button>SIGN OUT</button></div></div><div><h2>NOT FOUND: 404</h2><h3>THE PAGE YOU REQUESTED IS NOT AVAILABLE</h3><a href=\\"/\\">GO TO HOME PAGE</a></div></div>"`;
